Added Aiken (#1)
Riley-Kilgore/learnxinyminutes-docs: aiken.html.markdown

The language includes features tailored for blockchain operations and smart contract development. These include built-in functions based on what's available in UPLC, with upcoming additions like BLS12-381 curve primitives.

Aiken comes with a standard library and prelude that are particularly useful for creating Cardano smart contracts. The language supports various programming constructs such as functions, validators, pattern matching, and algebraic data types.

Aiken accomodates to developer experience with innovations like CIP-57 (Blueprints) having been implemented. This allows developers to write less code to build transactions that are compliant with their validators.

As a relatively new language, Aiken is evolving to meet the needs of the Cardano ecosystem. Its development is guided by the requirements of smart contract creation on the Cardano Blockchain.

```aiken
//// Basic Syntax and Structure

// Comments in Aiken start with '//'
// Module-level comments use '////'
// Documentation comments use '///'
//// This is a module-level comment
/// This is a documentation comment

// Constants are defined using 'const'

> const my_constant: Int = 42

Reviewer: ```suggestion
const my_int_constant: Int = 42
const my_bytearray_constant: ByteArray = "foo"
```

> [!NOTE]
>
> It would be nice to start with data-types before constants (like we do in the language tour) as everything is built on top of data-types.

// Functions are defined using 'fn'
fn add(a: Int, b: Int) -> Int {
  a + b
}

> // Public functions are marked with 'pub'

Reviewer: ```suggestion
// Functions are private by default. They can be exported with the 'pub' keyword.
```

pub fn public_function(x: Int) -> Int {
  x * 2
}

> 

Reviewer: It would be nice to also add illustrations for:

- Function capturing
- Pipeline
- Backpassing

//// Data Types

// Aiken has several built-in types:

> const my_int: Int = 10
> const my_string: String = @"Hello, Aiken!"
> const my_bool: Bool = True
> 
> // Lists
> const my_list: List<Int> = [1, 2, 3, 4, 5]
> 
> // Tuples
> const my_tuple: (Int, String) = (1, @"one")

Reviewer: ```suggestion
let my_int: Int = 10
let my_bytearray_utf8: ByteArray = "Hello, Aiken!"
let my_bytearray_hex: ByteArray = #"48656c6c6f2c2041696b656e21"
let my_bytearray_array: ByteArray = #[0x48, 0x65, 0x6c, 0x6c, 0x6f]
let my_string: String = @"Hello, Aiken!"
let my_bool: Bool = True

// Lists
let my_list: List<Int> = [1, 2, 3, 4, 5]

// Tuples
let my_tuple: (Int, ByteArray) = (1, "one")
```

> // Custom types can be defined using 'type'

Reviewer: ```suggestion
// Records can be defined using 'type', a constructor and typed fields. 
```

type RGB {
  red: Int,
  green: Int,
  blue: Int,

> }

Reviewer: We should show an example of record update syntax.

> 

Reviewer: ```suggestion

// Enums are also supported
type Season {
  Summer
  Fall
  Winter
  Spring
}
```

> // Enums are also supported

Reviewer: ```suggestion
// As well as full algebraic data-types, with (or without) generic arguments. 
```

Reviewer: Note: that isn't an enum 😅 ...

pub type Option<a> {
  Some(a)
  None
}

> 

Reviewer: ```suggestion

// Types can also be aliased
type CartesianCoordinates = (Int, Int)

```

// Validators are special functions in Aiken for smart contracts
validator {
  fn spend(datum: Data, redeemer: Data, context: Data) -> Bool {
    // Smart contract logic goes here
    True
  }

> }

Reviewer: Weird to have this one here and not alongside 'functions'

> //// Control Flow

Reviewer: `expect` and `error` are missing.

// If expressions
fn abs(x: Int) -> Int {
  if x < 0 {
    -x
  } else {
    x
  }
}

// When expressions (pattern matching)
fn describe_list(list: List<Int>) -> String {
  when list is {
    [] -> @"The list is empty"
    [x] -> @"The list has one element"
    [x, y, ..] -> @"The list has multiple elements"
  }

> }

Reviewer: ```suggestion
}

// 'and' and 'or' boolean operators come with their own syntactic sugar
fn my_validation_logic() -> Bool {
  (should_satisfy_condition_1 && should_satisfy_condition_2) || should_satisfy_condition_3 || should_satisfy_condition_4 
}

fn my_more_readable_validation_logic() -> Bool {
  or {
    and {
      should_satisfy_condition_1,
      should_satisfy_condition_2,   
    },
    should_satisfy_condition_3,
    should_satisfy_condition_4,
  }
}
```

// Trace for debugging
fn debug_function(x: Int) -> Int {
  trace @"We can trace using this syntax."
  (x * 2 < 10)? // This will trace if false.
}

//// Modules and Imports

> // Importing modules

Reviewer: ```suggestion
// Importing a qualified module
```

> use aiken/list

Reviewer: ```suggestion
use aiken/list
// Importing identifiers explicitly 
use aiken/math.{pow2}
```

// Using imported functions
fn use_imports() {
  let numbers = [1, 2, 3, 4, 5]

>   list.at(2) // evaluates to 3

Reviewer: ```suggestion
  pow2(list.at(2)) // evaluates to 8
```

}

//// Testing

// Tests are defined using the 'test' keyword
test positive_add() {
  let result = add(2, 3)
  result == 5

> }

Reviewer: ```suggestion
}

// Tests can be marked as expected to 'fail'
test not_true() fail {
  1 + 1 == 3
}
```

// Byte arrays
const my_bytes: ByteArray = #[01, 02, 03, 04]

> const my_bytes_2: ByteArray = #"1234"

Reviewer: ```suggestion
```

Reviewer: Moved earlier.

// Todo for unfinished code
fn unfinished_function() {
  todo

> }

Reviewer: ```suggestion
}

// Todo can take an optional (String) message
fn unfinished_function() {
  todo @"unimplemented: should return some fancy results" 
}
```

Aiken accomodates to developer experience with innovations like CIP-57 (Blueprints) having been implemented. This allows developers to write less code to build transactions that are compliant with their validators.

As a relatively new language, Aiken is evolving to meet the needs of the Cardano ecosystem. Its development is guided by the requirements of smart contract creation on the Cardano Blockchain.

```aiken
//// Variables & Constants

// Constants are defined using 'const'
// Constants may be used at the top level of a module
const my_constant: Int = 42

// Values assigned to let-bindings are immutable
// New bindings can shadow previous bindings
let x = 1
let y = x
let x = 2

// y + x == 3

//// Functions

// Functions are defined using 'fn'

// Functions can be named
fn add(a: Int, b: Int) -> Int {
  a + b
}

// Annotations are often optional (albeit recommended), as they can be inferred.
fn add_inferred(a, b) {
  a + b
}

// Functions are private by default. They can be exported with the 'pub' keyword.
pub fn public_function(x: Int) -> Int {
  x * 2
}

// Functions can accept functions as arguments

> fn apply_function_twice(f: fn(t) -> t, x) {

Reviewer: ```suggestion
fn apply_function_twice(x, f: fn(t) -> t) {
```

Reviewer: Passing the callback last, so it plays nice with backpassing :)

  f(f(x))
}

// We can achieve the same outcome as above with pipelining using '|>'

> fn apply_function_with_pipe_twice(f: fn(t) -> t, x) {

Reviewer: ```suggestion
fn apply_function_with_pipe_twice(x, f: fn(t) -> t) {
```

  x
  |> f
  |> f
}

> // Function capturing is demonstrated here using add to create add_one
> // We choose where the argument to be passed later with '_'.
> fn capture_demonstration() {
>   let add_one = add(_, 1)
> 
>   apply_function_twice(add_one, 2) // Evaluates to 4
> }

Reviewer: ```suggestion
// Function can be partially applied to create new functions. This behavior is also referred to as 'function captures'.
fn capture_demonstration() {
  let add_one = add(_, 1)
  apply_function_twice(add_one, 2) // Evaluates to 4
}
```

// Backpassing

> // The simplest example provided by the documentation is as follows
> let x <- foo()
>  
> // which is equivalent to writing:
> foo(fn(x) { })

Reviewer: ```suggestion
// Functions that take callbacks can leverage backpassing using `<-`
fn cubed_backpassing(n) {
  let total <- apply_function_twice(n)
  total * n
} 
    
// which is equivalent to writing:
fn cubed_callback(n) {
  apply_function_twice(n, fn(total) {
    total * n
  })
}
```

// A more in depth example follows.

// We define the following set of functions over 'Fuzzer'.
fn constant(a) -> Fuzzer<a>
fn bool() -> Fuzzer<Bool>
fn map(Fuzzer<a>, fn(a) -> b) -> Fuzzer<b>
fn and_then(Fuzzer<a>, fn(a) -> Fuzzer<b>) -> Fuzzer<b>

// To leverage them without backpassing is quite a syntactic mess
pub fn option(fuzz_a: Fuzzer<a>) -> Fuzzer<Option<a>> {
  bool()
    |> and_then(
         fn(predicate) {
           if predicate {
             fuzz_a |> map(Some)
           } else {
             constant(None)
           }
         },
       )
}

// Whereas with backpassing the same logic may be expressed more cleanly.
pub fn option(fuzz_a: Fuzzer<a>) -> Fuzzer<Option<a>> {
  let predicate <- and_then(bool())
  if predicate {
   fuzz_a |> map(Some)
  } else {
    constant(None)
  }

> }

Reviewer: Maybe fuzzers are a little too advanced / off-topic for this. Especially since they do not introduce new syntax elements (with the exception of backpassing perhaps, but it's covered earlier already).

  RGB {..rgb, red: 255}
}

// Enums are also supported
// As well as full algebraic data-types, with (or without) generic arguments. 
pub type Option<a> {
  Some(a)
  None
}

// Validators are special functions in Aiken for smart contracts
validator {
  fn spend(datum: Data, redeemer: Data, context: Data) -> Bool {
    // Smart contract logic goes here
    True
  }
}

// 'Data' can be casted to an extracted from using custom types.
fn to_datum(datum: Data) -> Datum {
    expect d: Datum = datum
    // The line above will fail if datum is not a valid representation of Datum.
    d
}

// Custom data types may be used in contexts that require 'Data'
// The custom data type will be implicitly converted to 'Data'

//// Control Flow

// If expressions
fn abs(x: Int) -> Int {
  if x < 0 {
    -x
  } else {
    x
  }
}

// When expressions (pattern matching)
fn describe_list(list: List<Int>) -> String {
  when list is {
    [] -> @"The list is empty"
    [x] -> @"The list has one element"
    [x, y, ..] -> @"The list has multiple elements"
  }
}

// This syntax can be used instead of when to check a single case and fail otherwise.
expect Some(foo) = my_optional_value

> // Using 'expect' to check conditions

Reviewer: ```suggestion
// When the left-hand side pattern is `True`, it can be omitted for conciseness. 
```
